README.md
# brownpaperbag

[![image](https://img.shields.io/pypi/v/brownpaperbag.svg)](https://pypi.python.org/pypi/brownpaperbag)
![GitHub](https://img.shields.io/github/license/bpaulin/brownpaperbag.svg)
![PyPI - Downloads](https://img.shields.io/pypi/dm/brownpaperbag.svg)

[![image](https://img.shields.io/travis/bpaulin/brownpaperbag.svg)](https://travis-ci.org/bpaulin/brownpaperbag)
[![Updates](https://pyup.io/repos/github/bpaulin/brownpaperbag/shield.svg)](https://pyup.io/repos/github/bpaulin/brownpaperbag/)


> But you try and tell the young people today that...

## installation

```bash
pip install brownpaperbag
```

## Usage (cli)

### Configuration

You need to use the official app, and to be logged with the installer code

You'll need the open password: From the app, you can set it on Other>System.
You'll also need your gateway's IP: Again from the app, you can find it in Other>Network

So if your ip is 192.168.1.10 and your password is qwerty:

```bash
export BPB_HOST=192.168.1.10
export BPB_PASSWORD=qwerty
export BPB_PORT=20000
```

### Discovering devices

the command ```brownpaperbag event --human``` can help you to discover devices

```bash
$ brownpaperbag event --human
light 15 is OFF (*1*0*15##)
light 11 is OFF (*1*0*11##)
light 0010 is OFF (*1*0*0010##)
light 0011 is OFF (*1*0*0011##)
light 0012 is OFF (*1*0*0012##)
light 02 is OFF (*1*0*02##)
light 0014 is OFF (*1*0*0014##)
light 0111 is OFF (*1*0*0111##)
light 04 is OFF (*1*0*04##)
light 19 is OFF (*1*0*19##)
light 01 is OFF (*1*0*01##)
light 09 is OFF (*1*0*09##)
cover 07 is STOPPED (*2*0*07##)
cover 0113 is STOPPED (*2*0*0113##)
cover 13 is STOPPED (*2*0*13##)
cover 17 is STOPPED (*2*0*17##)
cover 16 is STOPPED (*2*0*16##)
cover 06 is STOPPED (*2*0*06##)
cover 08 is STOPPED (*2*0*08##)
cover 06 is UP (*2*1000#1*06##)
cover 06 is STOPPED (*2*1000#0*06##)
cover 06 is DOWN (*2*1000#2*06##)
cover 06 is STOPPED (*2*1000#0*06##)
light 0111 is ON (*1*1*0111##)
light 0111 is OFF (*1*0*0111##)
light 0111 is ON (*1*1*0111##)
light 0111 is OFF (*1*0*0111##)
cover 06 is DOWN (*2*1000#2*06##)
cover 06 is STOPPED (*2*1000#0*06##)
cover 06 is UP (*2*1000#1*06##)
cover 06 is STOPPED (*2*0*06##)

```


HISTORY.md
=======
History
=======

0.0.5 (2019-02-15)
------------------

* First release on PyPI.


brownpaperbag/cli.py
# -*- coding: utf-8 -*-

"""Console script for brownpaperbag."""
import sys
import click
import asyncio
import re
import logging
from brownpaperbag import BpbGate, SESSION_EVENT


@click.group()
@click.option("--host", envvar="BPB_HOST", prompt=True, help="MyHomeServer1 IP")
@click.option(
    "--port", envvar="BPB_PORT", default=20000, prompt=True, help="MyHomeServer1 port"
)
@click.option(
    "--password",
    envvar="BPB_PASSWORD",
    prompt=True,
    hide_input=True,
    help="OPEN password",
)
@click.option("--verbose", is_flag=True)
@click.pass_context
def main(ctx, host, port, password, verbose):
    """Console script for brownpaperbag.

    Provides interaction with myhomeserver1
    """
    gate = BpbGate(host, port, password)
    gate.logger = logging.basicConfig(level=logging.DEBUG)
    ctx.ensure_object(dict)
    ctx.obj["GATE"] = gate
    return 0


@main.command()
@click.option("--human", is_flag=True, help="human readable")
@click.pass_context
def event(ctx, human):
    """Subscribe to gateway events."""
    gate = ctx.obj["GATE"]
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(gate.command_session(SESSION_EVENT))
    except (ConnectionError, NotImplementedError) as e:
        click.secho(str(e), fg="red")
        return 1

    statuses = {}
    while True:
        event = loop.run_until_complete(gate.readevent())
        if not human:
            click.echo(event)
        else:
            (who, what, where) = re.search(r"\*(.*)\*(.*)\*(.*)##", event).groups()
            if what.startswith("1000#"):
                what = what.replace("1000#", "")
            if who not in statuses.keys():
                statuses[who] = {}
            if where not in statuses[who].keys() or what != statuses[who][where]:
                statuses[who][where] = what
                if who == "1":
                    status = "OFF"
                    if what == "1":
                        status = "ON"
                    click.echo("light %s is %s (%s)" % (where, status, event))
                if who == "2":
                    status = "STOPPED"
                    if what == "1":
                        status = "UP"
                    elif what == "2":
                        status = "DOWN"
                    click.echo("cover %s is %s (%s)" % (where, status, event))


@main.command()
@click.argument("command")
@click.pass_context
def raw(ctx, command):
    """Send raw openwebnet command."""
    gate = ctx.obj["GATE"]
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(gate.command_session())
    except (ConnectionError, NotImplementedError) as e:
        click.secho(str(e), fg="red")
        return 1

    click.echo(loop.run_until_complete(gate.send_raw(command)))


@main.command()
@click.option(
    "--status", "operation", flag_value="status", default=True, help="get status"
)
@click.option("--on", "operation", flag_value="on", help="Turn On")
@click.option("--off", "operation", flag_value="off", help="Turn Off")
@click.argument("id")
@click.pass_context
def light(ctx, operation, id):
    """Interact with a light."""
    gate = ctx.obj["GATE"]
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(gate.command_session())
    except (ConnectionError, NotImplementedError) as e:
        click.secho(str(e), fg="red")
        return 1
    if operation == "on":
        loop.run_until_complete(gate.turn_on_light(id))
    elif operation == "off":
        loop.run_until_complete(gate.turn_off_light(id))

    if loop.run_until_complete(gate.is_light_on(id)):
        click.echo("ON")
    else:
        click.echo("OFF")


@main.command()
@click.option(
    "--status", "operation", flag_value="status", default=True, help="get status"
)
@click.option("--up", "operation", flag_value="Open")
@click.option("--down", "operation", flag_value="Close")
@click.option("--stop", "operation", flag_value="Stop")
@click.argument("id")
@click.pass_context
def cover(ctx, operation, id):
    """Interact with a cover."""
    gate = ctx.obj["GATE"]
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(gate.command_session())
    except (ConnectionError, NotImplementedError) as e:
        click.secho(str(e), fg="red")
        return 1
    if operation == "up":
        loop.run_until_complete(gate.open_cover(id))
    elif operation == "down":
        loop.run_until_complete(gate.close_cover(id))
    elif operation == "stop":
        loop.run_until_complete(gate.stop_cover(id))

    click.echo(loop.run_until_complete(gate.get_cover_state(id)))


@main.command()
@click.option("--lights/--no-lights", default=True, help="Include Lights")
@click.option("--covers/--no-covers", default=True, help="Include Covers")
@click.pass_context
def list(ctx, lights, covers):
    """List known devices."""
    gate = ctx.obj["GATE"]
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(gate.command_session())
    except (ConnectionError, NotImplementedError) as e:
        click.secho(str(e), fg="red")
        return 1
    if lights:
        click.echo("Lights: ")
        bpb_lights = loop.run_until_complete(gate.get_light_ids())
        for index, value in bpb_lights.items():
            click.echo("%s: %s" % (index.rjust(5), value))
    if covers:
        click.echo("Covers: ")
        bpb_covers = loop.run_until_complete(gate.get_cover_ids())
        for index, value in bpb_covers.items():
            click.echo("%s: %s" % (index.rjust(5), value))


if __name__ == "__main__":
    sys.exit(main(auto_envvar_prefix="BPB", obj={}))  # pragma: no cover
